File: jvm/sargon-android/src/main/java/com/radixdlt/sargon/extensions/Blob.kt
package com.radixdlt.sargon.extensions

import com.radixdlt.sargon.BagOfBytes
import com.radixdlt.sargon.Blob
import com.radixdlt.sargon.Blobs
import com.radixdlt.sargon.blobToBytes
import com.radixdlt.sargon.blobToString
import com.radixdlt.sargon.blobsListOfBlobs
import com.radixdlt.sargon.newBlobFromBytes
import com.radixdlt.sargon.newBlobsFromBlobList

fun Blob.Companion.init(bytes: BagOfBytes) = newBlobFromBytes(bytes = bytes)

val Blob.bytes: BagOfBytes
    get() = blobToBytes(blob = this)

val Blob.string: String
    get() = blobToString(blob = this)

fun Blobs.Companion.init(blobs: List<Blob>) = newBlobsFromBlobList(blobs = blobs)

fun Blobs.Companion.init(vararg blob: Blob) = init(blobs = blob.asList())

fun Blobs.toList(): List<Blob> = blobsListOfBlobs(blobs = this)

File: crates/sargon-uniffi/src/wrapped_radix_engine_toolkit/low_level/transaction_manifest/blobs/blob.rs
use crate::prelude::*;
use sargon::BagOfBytes as InternalBagOfBytes;
use sargon::Blob as InternalBlob;

/// Blob is a wrapper a bag of bytes
#[derive(Clone, PartialEq, Eq, uniffi::Record)]
pub struct Blob {
    pub secret_magic: BagOfBytes,
}

impl Blob {
    pub fn into_internal(&self) -> InternalBlob {
        self.clone().into()
    }
}

impl From<InternalBlob> for Blob {
    fn from(internal: InternalBlob) -> Self {
        Self {
            secret_magic: internal.0.into(),
        }
    }
}

impl Into<InternalBlob> for Blob {
    fn into(self) -> InternalBlob {
        InternalBlob(self.secret_magic.into())
    }
}

#[uniffi::export]
pub fn new_blob_from_bytes(bytes: BagOfBytes) -> Blob {
    InternalBlob::from(bytes.into_internal()).into()
}

#[uniffi::export]
pub fn blob_to_bytes(blob: &Blob) -> BagOfBytes {
    blob.secret_magic.clone()
}

#[uniffi::export]
pub fn blob_to_string(blob: &Blob) -> String {
    blob.into_internal().to_string()
}

decl_conversion_tests_for!(Blob);
